backend/controllers/user.controller.ts

```diff
@@ -0,0 +1,122 @@
+import {User} from '../models/user.models';
+import * as bcrypt from 'bcryptjs';
+import * as jwt from 'jsonwebtoken';
+import * as fs from 'fs';
+import {databaseSecret} from '../environment';
+
+const generateAuthToken = async (user: any) => {
+  const token = jwt.sign({_id: user._id}, databaseSecret);
+  user.tokens = user.tokens.concat({token});
+  await user.save();
+  return token;
+};
+
+const findByCredentials = async (username: string, password: string) => {
+  // Search for a user by username and password.
+  const user = await User.findOne({username});
+  if (!user) {
+    throw new Error('AUTH_FAIL');
+  }
+  // @ts-ignore
+  const isPasswordMatch = await bcrypt.compare(password, user.password);
+  if (!isPasswordMatch) {
+    throw new Error('AUTH_FAIL');
+  }
+  return user;
+};
+
+export const registerUser = async (req: any, res: any) => {
+  try {
+    // Req.body should have username and password. isAuthor also doubles as staff for this application
+    const data = {
+      firstName: req.body.firstName,
+      lastName: req.body.lastName,
+      username: req.body.username,
+      password: req.body.password
+    };
+    if (data.password.length < 6) {
```

Comment:
Yes. Each error message can be treated like a code. For example, if the backend returns 'AUTH_FAIL', then in 'auth.service', there is a function along the lines of "handleErrors". This will take this "AUTH_FAIL" code and return an error message for angular to deliver to the frontend for the user to read.